kikuchipy/indexing/_refinement.py

```diff
@@ -0,0 +1,1686 @@
+    phi2 = r[..., 2]
+
+    args = (exp,) + pre_args + (dc,)
+
+    r_x0 = np.array((phi1, Phi, phi2), dtype=np.float32)
+
+    if method.__name__ == "minimize":
+        soln = method(
+            _orientation_objective_function_euler,
+            x0=r_x0,
+            args=args,
+            **method_kwargs,
+        )
+    elif method.__name__ == "differential_evolution":
+        soln = method(
+            _orientation_objective_function_euler,
+            bounds=[
+                (r_x0[0] - trust_region[0], r_x0[0] + trust_region[0]),
+                (r_x0[1] - trust_region[1], r_x0[1] + trust_region[1]),
+                (r_x0[2] - trust_region[2], r_x0[2] + trust_region[2]),
+            ],
+            args=args,
+            **method_kwargs,
+        )
+    elif method.__name__ == "dual_annealing":
+        soln = method(
+            _orientation_objective_function_euler,
+            bounds=[
+                (r_x0[0] - trust_region[0], r_x0[0] + trust_region[0]),
+                (r_x0[1] - trust_region[1], r_x0[1] + trust_region[1]),
+                (r_x0[2] - trust_region[2], r_x0[2] + trust_region[2]),
+            ],
+            args=args,
+            **method_kwargs,
+        )
+    elif method.__name__ == "basinhopping":
+        method_kwargs["minimizer_kwargs"]["args"] = args
+        soln = method(
+            _orientation_objective_function_euler,
+            x0=r_x0,
+            **method_kwargs,
+        )
+
+    score = 1 - soln.fun
+    refined_phi1 = soln.x[0]
+    refined_Phi = soln.x[1]
+    refined_phi2 = soln.x[2]
+
+    return (score, refined_phi1, refined_Phi, refined_phi2)
+
+
+# SIMILARITY METRICS
+@numba.njit(fastmath=True)
+def _py_ncc(a: np.ndarray, b: np.ndarray) -> float:
+    """Calculates the normalized cross-correlation coefficient (NCC)
+    between two patterns of the same size and with data type float32.
+
+    Parameters
+    ----------
+    a, b
+        np.ndarray with shape (n x m) and with data type np.float32.
+
+    Returns
+    -------
+        float representing the NCC between a and b.
+    """
+    abar = np.mean(a)
+    bbar = np.mean(b)
+    astar = a - abar
+    bstar = b - bbar
+    return np.sum(astar * bstar) / np.sqrt(
+        np.sum(np.square(astar)) * np.sum(np.square(bstar))
+    )
+
+
+def _refinement_parameter_check(exp, xmap, detector, method, mask):
+    """Helper function to determine if the input to the different EBSD
+    refinement methods are valid.
+
+    Parameters
+    ----------
+    exp
+        Experimental data
+    xmap
+        CrystalMap containing the indexing result.
+    detector
+        EBSDDetector describing the experimental geometry.
+    method
+        Name of the scipy optimization method.
+    mask
+        Mask applied to the patterns.
+
+    Raises
+    ------
+    ValueError
+        If one of the parameters are invalid.
+    """
+    # Signal and Detector must have same shape
+    if exp.axes_manager.signal_shape != detector.shape:
```

Comment:
signal_shape[::-1] must be used, since HyperSpy uses (x, y) while EBSDDetector uses (y, x).